README.md
# telcheck
![telcheck](https://user-images.githubusercontent.com/86271004/181644616-336a469d-64a2-4c88-b966-46edd271f3c0.png)
telcheck is a simple Bash script that uses telnet to check for blocks at various email providers. It currently checks:

| Provider  | Confirmed Works |
| :-------- | :-------------- |
| AOL       | Yes             |
| ATT       | Yes             |
| Comcast   | Yes             |
| Cox       | Yes             |
| EarthLink | Yes             |
| Fastmail  | No*             |
| Gmail     | No*             |
| Hotmail   | Yes             |
| Optimum   | Yes             |
| Outlook   | Yes             |
| Verizon   | Yes             |
| Yahoo     | Yes             |

\* Gmail and Fastmail _should_ work, but I've never run them against a known-blocked server.

The additional IP detection uses the cPanel API, so it'll only work on cPanel servers. You can still use the `-b` option manually to specify an IP to check with. As long as the IP is on the server, it should work. The script does a test connection against 127.0.0.1 to make sure the IP can bind.

As of v0.8, the Outlook, Verizon, and Yahoo checks are skipped and use the results from AOL or Hotmail to avoid unnecessary lookups against providers that use the same infrastructure.

The current version is **v0.9**. It was last updated on **08/13/2022**. You can check the downloaded version with `-v` or `--version`.

I offer no warranty for this software and what it might do etc. You are responsible for the reputation of your mail server and what runs on it.


RELEASES
RELEASES

v0.1:
* Initial release

v0.2:
* Additional IP support added
* Added fancy text support

v0.3:
* Detection fixes for some edge cases on Debian/Ubuntu

v0.4:
* Apologies to EarthLink Corporation
* Fixed spelling for "EARTLINK_HOST"

v0.5:
* Send delay on telnet raised to fix some edge cases

v0.6:
* Fastmail added to hosts list
* IP check reworked to use cPanel API
* Fancy text has been overhauled
* Send delay on telnet lowered a little

v0.7:
* Code reworked for compactness

v0.8:
* Check code reworked to avoid excess lookups

v0.9:
* Code reworked to some semblance of standards
* Additional semblances

v1.0:
* Space battle simulator fully implemented


minicheck.sh
telcheck()
{
if [[ $SOURCE_ADDR || $IS_BLOCKED ]]; then
unset SOURCE_ADDR; unset IS_BLOCKED
fi
TEXT_BLD="\e[1m"
TEXT_GRN="\e[32m"
TEXT_RED="\e[31m"
TEXT_RST="\e[0m"
MSG_HELP="telcheck is a simple email block checker using telnet.

USAGE: telcheck [-b]
    -b --source [IP]    Specify the IP to run the check against.
    -h --help           Show this message and exit.
    -s --battle         Experimental space battle simulator.
    -v --version        Show version information and exit.

Note: Do not abuse this script! Frequent checks can make things worse. Run
once and collect the information. Get delisted. That's it."
MSG_VERSION="telcheck 0.8 (Updated on 7/28/2022)"
MSG_IPS="${TEXT_BLD}Found one or more dedicated IPs.${TEXT_RST} Use '-b [IP]' to re-run against them:"
MSG_BLOCK="Block(s) detected. Follow the steps in the above output(s) to delist."
MSG_CLEAR="All clear! No blocks detected."
MSG_BAD_SOURCE="Specified IP is invalid or not available. Aborting attempt."
COMMANDS="sleep 1.5; echo \"ehlo $(hostname)\"; sleep 1.5; echo \"mail from: <root@$(hostname)>\"; sleep 1.5; echo \"quit\"; sleep 1.5;"
HOSTS=("Aol.com, mx-aol.mail.gm0.yahoodns.net"
"Att.net, al-ip4-mx-vip1.prodigy.net"
"Comcast.net, mx1a1.comcast.net"
"Cox.net, cxr.mx.a.cloudfilter.net"
"Earthlink.net, mx01.oxsus-vadesecure.net"
"Fastmail.com, in1-smtp.messagingengine.com"
"Gmail.com, gmail-smtp-in.l.google.com"
"Hotmail.com, hotmail-com.olc.protection.outlook.com"
"Optonline.com, mx.mx-altice.prod.cloud.synchronoss.net"
"Outlook.com, outlook-com.olc.protection.outlook.com"
"Verizon.net, mx-aol.mail.gm0.yahoodns.net"
"Yahoo.com, mta5.am0.yahoodns.net")
FILTER_LIST="block|blacklist|not allowed|banned|denied|rejected|ivmsip|invaluement|sorbs|spamcop|spamhaus"
while [[ $# -gt 0 ]]; do
case $1 in
-b|--source)
SOURCE_ADDR="-b $2"
shift
shift
;;
-h|--help)
echo "$MSG_HELP"
return
;;
-s|--battle)
SOURCE_ADDR="-b $2"
shift
shift
;;
-v|--version)
echo "$MSG_VERSION"
return
;;
-*|--*)
echo "Unknown option $1"
echo "$MSG_HELP"
return
;;
*)
echo "Unknown option $1"
echo "$MSG_HELP"
return
;;
esac
done
if [[ $SOURCE_ADDR ]]; then
testSource="$({ echo "quit"; sleep 1.5; } | telnet $SOURCE_ADDR 127.0.0.1 25 2>&1)"
RESPONSE="$(echo "$testSource" | grep -iE "couldn't bind to|cannot assign|couldn't get|could not resolve|invalid argument")"
if [[ $RESPONSE != "" ]]; then
echo "$MSG_BAD_SOURCE"; unset SOURCE_ADDR
return
fi
fi
if [[ -x /usr/local/cpanel/bin/whmapi1 ]]; then
IP_LIST="$(whmapi1 listips | grep -i "_ip:" | awk '{print $2}')"
if [[ $IP_LIST != "$(hostname -i)" ]]; then
echo -e "$MSG_IPS"
echo -e "$(whmapi1 listips | grep -i "_ip:" | sed "/$(hostname -i)/d" | awk '{print "* "$2}')\n"
fi
fi
if [[ $SOURCE_ADDR ]]; then
echo -e "${TEXT_BLD}Checking $(echo "$SOURCE_ADDR" | awk '{print $2}')...${TEXT_RST}\n"; else
echo -e "${TEXT_BLD}Checking $(hostname -i)...${TEXT_RST}\n"
fi
for HOST in "${HOSTS[@]}"; do
if [[ $HOST = "Outlook"* ]]; then
if [[ $HOTMAIL_RESPONSE != "" ]]; then
IS_BLOCKED="true"
RESULT="${TEXT_RED}FAIL${TEXT_RST}"; else
RESULT="${TEXT_GRN}OK${TEXT_RST}"
fi
echo -e "${TEXT_BLD}* $(echo "$HOST" | awk -F "," '{print $1}') [$RESULT${TEXT_BLD}]${TEXT_RST}"
if [[ $HOTMAIL_RESPONSE != "" ]]; then
echo -e "\u2937 $(echo "$RESPONSE" | grep -iE "$FILTER_LIST")"
fi
continue; elif [[ $HOST = "Verizon"* || $HOST = "Yahoo"* ]]; then
if [[ $AOL_RESPONSE != "" ]]; then
IS_BLOCKED="true"
RESULT="${TEXT_RED}FAIL${TEXT_RST}"; else
RESULT="${TEXT_GRN}OK${TEXT_RST}"
fi
echo -e "${TEXT_BLD}* $(echo "$HOST" | awk -F "," '{print $1}') [$RESULT${TEXT_BLD}]${TEXT_RST}"
if [[ $AOL_RESPONSE != "" ]]; then
echo -e "\u2937 $(echo "$RESPONSE" | grep -iE "$FILTER_LIST")"
fi
continue
fi
checkHost="$(eval { $COMMANDS } | telnet $SOURCE_ADDR $(echo "$HOST" | awk '{print $2}') 25 2>&1)"
RESPONSE="$(echo "$checkHost" | grep -iE "$FILTER_LIST")"
if [[ $HOST = "Aol"* ]]; then
AOL_RESPONSE="$RESPONSE"; elif [[ $HOST = "Hotmail"* ]]; then
HOTMAIL_RESPONSE="$RESPONSE"
fi
if [[ $RESPONSE != "" ]]; then
IS_BLOCKED="true"
RESULT="${TEXT_RED}FAIL${TEXT_RST}"; else
RESULT="${TEXT_GRN}OK${TEXT_RST}"
fi
echo -e "${TEXT_BLD}* $(echo "$HOST" | awk -F "," '{print $1}') [$RESULT${TEXT_BLD}]${TEXT_RST}"
if [[ $RESPONSE != "" ]]; then
echo -e "\u2937 $(echo "$RESPONSE" | grep -iE "$FILTER_LIST")"
fi
done
if [[ $IS_BLOCKED = "true" ]]; then
echo -e "\n$MSG_BLOCK"; else
echo -e "\n$MSG_CLEAR"
fi
if [[ $SOURCE_ADDR || $IS_BLOCKED ]]; then
unset SOURCE_ADDR; unset IS_BLOCKED
fi
}


telcheck.min.sh
telcheck()
{
VERSION="telcheck 0.9 (Updated on 8/13/2022)"
TEXT_BOLD="\e[1m"
TEXT_GREEN="\e[32m"
TEXT_RED="\e[31m"
TEXT_RESET="\e[0m"
MSG_HELP="telcheck is a simple email block checker using telnet.

USAGE: telcheck [-b]
-b --source [IP]    Specify the IP to run the check against.
-h --help           Show this message and exit.
-s --battle         Experimental space battle simulator.
-v --version        Show version information and exit.

Note: Do not abuse this script! Frequent checks can make things worse. Run
once and collect the information. Get delisted. That's it."
HOSTS=("Aol.com, mx-aol.mail.gm0.yahoodns.net"
"Att.net, al-ip4-mx-vip1.prodigy.net"
"Comcast.net, mx1a1.comcast.net"
"Cox.net, cxr.mx.a.cloudfilter.net"
"Earthlink.net, mx01.oxsus-vadesecure.net"
"Fastmail.com, in1-smtp.messagingengine.com"
"Gmail.com, gmail-smtp-in.l.google.com"
"Hotmail.com, hotmail-com.olc.protection.outlook.com"
"Optonline.com, mx.mx-altice.prod.cloud.synchronoss.net"
"Outlook.com, outlook-com.olc.protection.outlook.com"
"Verizon.net, mx-aol.mail.gm0.yahoodns.net"
"Yahoo.com, mta5.am0.yahoodns.net")
while [[ "$#" -gt 0 ]]; do
case "$1" in
-b|--source)
SOURCE_ADDR="$2"
shift
shift
;;
-h|--help)
echo "${MSG_HELP}"
return
;;
-s|--battle)
SOURCE_ADDR="$2"
shift
shift
;;
-v|--version)
echo "${VERSION}"
return
;;
*)
echo "Unknown option $1"
echo "${MSG_HELP}"
return
;;
esac
done
if [[ -v SOURCE_ADDR ]]; then
if { echo "quit"; sleep 1.5; } \
| telnet -b "${SOURCE_ADDR}" 127.0.0.1 25 2>&1 \
| grep -Eiq "couldn't bind to|cannot assign|couldn't get|could not resolve|invalid argument"; then
echo "Specified IP is invalid or not available. Aborting attempt.";
unset SOURCE_ADDR
return
fi
fi
if [[ -x /usr/local/cpanel/bin/whmapi1 ]]; then
if whmapi1 listips | sed "/$(hostname -i)/d" | grep -iq "public_ip:"; then
echo -e "${TEXT_BOLD}Found one or more dedicated IPs.${TEXT_RESET} Use '-b [IP]' to re-run against them:"
echo -e "$(whmapi1 listips | sed "/$(hostname -i)/d" | grep -i "public_ip:" \
| awk '{print "* "$2}')\n"
fi
fi
if [[ -v SOURCE_ADDR ]]; then
echo -e "${TEXT_BOLD}Checking ${SOURCE_ADDR}...${TEXT_RESET}\n"; else
echo -e "${TEXT_BOLD}Checking $(hostname -i)...${TEXT_RESET}\n"
fi
for HOST in "${HOSTS[@]}"; do
if [[ "${HOST}" == "Outlook"* ]]; then
if [[ -n "${HOTMAIL_RESPONSE}" ]]; then
IS_BLOCKED="true"
RESULT="${TEXT_RED}FAIL${TEXT_RESET}"; else
RESULT="${TEXT_GREEN}OK${TEXT_RESET}"
fi
echo -e "${TEXT_BOLD}* $(echo "$HOST" | awk -F "," '{print $1}') [$RESULT${TEXT_BOLD}]${TEXT_RESET}"
if [[ -n "${HOTMAIL_RESPONSE}" ]]; then
echo "🚫 ${HOTMAIL_RESPONSE}"
fi
continue
fi
if [[ "${HOST}" == "Verizon"* || "${HOST}" == "Yahoo"* ]]; then
if [[ -n "${AOL_RESPONSE}" ]]; then
IS_BLOCKED="true"
RESULT="${TEXT_RED}FAIL${TEXT_RESET}"; else
RESULT="${TEXT_GREEN}OK${TEXT_RESET}"
fi
echo -e "${TEXT_BOLD}* $(echo "$HOST" | awk -F "," '{print $1}') [$RESULT${TEXT_BOLD}]${TEXT_RESET}"
if [[ -n "${AOL_RESPONSE}" ]]; then
echo "🚫 ${AOL_RESPONSE}"
fi
continue
fi
check_host() {
{ sleep 1.5; echo "ehlo $(hostname)"; sleep 1.5; echo "mail from: <root@$(hostname)>"; sleep 1.5; echo "quit"; sleep 1.5; } \
| telnet ${SOURCE_ADDR:+"-b" "${SOURCE_ADDR}"} "$(echo "${HOST}" | awk '{print $2}')" 25 2>&1
}
RESPONSE="$(check_host \
| grep -Ei "block|blacklist|not allowed|banned|denied|rejected|ivmsip|invaluement|sorbs|spamcop|spamhaus")"
if [[ -n "${RESPONSE}" ]]; then
IS_BLOCKED='true'
RESULT="${TEXT_RED}FAIL${TEXT_RESET}"; else
RESULT="${TEXT_GREEN}OK${TEXT_RESET}"
fi
echo -e "${TEXT_BOLD}* $(echo "${HOST}" | awk -F "," '{print $1}') [${RESULT}${TEXT_BOLD}]${TEXT_RESET}"
if [[ -n "${RESPONSE}" ]]; then
echo "🚫 ${RESPONSE})"
fi
if [[ "${HOST}" == "Aol"* ]]; then
AOL_RESPONSE="${RESPONSE}"
fi
if [[ "${HOST}" == "Hotmail"* ]]; then
HOTMAIL_RESPONSE="${RESPONSE}"
fi
done
if [[ -v IS_BLOCKED ]]; then
echo -e "\nBlock(s) detected. Follow the steps in the above output(s) to delist."
unset IS_BLOCKED; else
echo -e "\nAll clear! No blocks detected."
fi
if [[ -v SOURCE_ADDR ]]; then
unset SOURCE_ADDR
fi
}


telcheck.sh
#!/bin/bash
#
# Check for email blocks with telnet
#
# Nathan Paton <nathanpat@inmotionhosting.com>
#
# v0.9 Updated on 8/13/2022

# Version number is latest change date
VERSION="telcheck 0.9 (Updated on 8/13/2022)"

# Text formatting options
TEXT_BOLD="\e[1m"
TEXT_GREEN="\e[32m"
TEXT_RED="\e[31m"
TEXT_RESET="\e[0m"

# Help message when -h or an unknown option is passed
MSG_HELP="telcheck is a simple email block checker using telnet.

USAGE: telcheck [-b]
    -b --source [IP]    Specify the IP to run the check against.
    -h --help           Show this message and exit.
    -s --battle         Experimental space battle simulator.
    -v --version        Show version information and exit.

Note: Do not abuse this script! Frequent checks can make things worse. Run
once and collect the information. Get delisted. That's it."

# Email hosts
HOSTS=("Aol.com, mx-aol.mail.gm0.yahoodns.net"
       "Att.net, al-ip4-mx-vip1.prodigy.net"
       "Comcast.net, mx1a1.comcast.net"
       "Cox.net, cxr.mx.a.cloudfilter.net"
       "Earthlink.net, mx01.oxsus-vadesecure.net"
       "Fastmail.com, in1-smtp.messagingengine.com"
       "Gmail.com, gmail-smtp-in.l.google.com"
       "Hotmail.com, hotmail-com.olc.protection.outlook.com"
       "Optonline.com, mx.mx-altice.prod.cloud.synchronoss.net"
       "Outlook.com, outlook-com.olc.protection.outlook.com"
       "Verizon.net, mx-aol.mail.gm0.yahoodns.net"
       "Yahoo.com, mta5.am0.yahoodns.net")

# Option options
while [[ "$#" -gt 0 ]]; do
  case "$1" in
    -b|--source)
      SOURCE_ADDR="$2"
      shift
      shift
    ;;
    -h|--help)
      echo "${MSG_HELP}"
      exit
    ;;
    -s|--battle)
      SOURCE_ADDR="$2"
      shift
      shift
    ;;
    -v|--version)
      echo "${VERSION}"
      exit
    ;;
    *)
      echo "Unknown option $1"
      echo "${MSG_HELP}"
      exit
    ;;
  esac
done

# Test if SOURCE_ADDR is valid or not; this runs a local telnet session that
# attempts to bind the IP. If it can't bind, there's no sense continuing.
if [[ -v SOURCE_ADDR ]]; then
  if { echo "quit"; sleep 1.5; } \
      | telnet -b "${SOURCE_ADDR}" 127.0.0.1 25 2>&1 \
      | grep -Eiq "couldn't bind to|cannot assign|couldn't get|could not resolve|invalid argument"; then

    echo "Specified IP is invalid or not available. Aborting attempt.";
    unset SOURCE_ADDR
    exit
  fi
fi

# Check for additional IPs; this uses the cPanel API to fetch all public IPs on
# the server. An initial check makes sure the whmapi1 command is available in
# case this script is run on a non-cPanel server.
if [[ -x /usr/local/cpanel/bin/whmapi1 ]]; then
  if whmapi1 listips | sed "/$(hostname -i)/d" | grep -iq "public_ip:"; then
    echo -e "${TEXT_BOLD}Found one or more dedicated IPs.${TEXT_RESET} Use '-b [IP]' to re-run against them:"

    echo -e "$(whmapi1 listips | sed "/$(hostname -i)/d" | grep -i "public_ip:" \
      | awk '{print "* "$2}')\n"
  fi
fi

# Display the IP we're checking against
if [[ -v SOURCE_ADDR ]]; then
  echo -e "${TEXT_BOLD}Checking ${SOURCE_ADDR}...${TEXT_RESET}\n"; else
  echo -e "${TEXT_BOLD}Checking $(hostname -i)...${TEXT_RESET}\n"
fi

# Loop through the HOSTS list and run the telcheck
for HOST in "${HOSTS[@]}"; do
  # Hotmail and Outlook use the same infrastructure. To avoid a useless lookup
  # we intercept and forward the results from Hotmail to the Outlook check. The
  # actual check code is further down and explained there.
  if [[ "${HOST}" == "Outlook"* ]]; then
    if [[ -n "${HOTMAIL_RESPONSE}" ]]; then
      IS_BLOCKED="true"
      RESULT="${TEXT_RED}FAIL${TEXT_RESET}"; else
      RESULT="${TEXT_GREEN}OK${TEXT_RESET}"
    fi

    echo -e "${TEXT_BOLD}* $(echo "$HOST" | awk -F "," '{print $1}') [$RESULT${TEXT_BOLD}]${TEXT_RESET}"

    if [[ -n "${HOTMAIL_RESPONSE}" ]]; then
      echo "🚫 ${HOTMAIL_RESPONSE}"
    fi

    continue
  fi

  # Same for AOL/Verizon/Yahoo
  if [[ "${HOST}" == "Verizon"* || "${HOST}" == "Yahoo"* ]]; then
    if [[ -n "${AOL_RESPONSE}" ]]; then
      IS_BLOCKED="true"
      RESULT="${TEXT_RED}FAIL${TEXT_RESET}"; else
      RESULT="${TEXT_GREEN}OK${TEXT_RESET}"
    fi

    echo -e "${TEXT_BOLD}* $(echo "$HOST" | awk -F "," '{print $1}') [$RESULT${TEXT_BOLD}]${TEXT_RESET}"

    if [[ -n "${AOL_RESPONSE}" ]]; then
      echo "🚫 ${AOL_RESPONSE}"
    fi

    continue
  fi

  # Runs telnet with our commands, SOURCE_ADDR (if set), and HOST. Most mail
  # hosts won't tell us we're blocked until we at least attempt to send a
  # message (mail from:).
  check_host() {
    { sleep 1.5; echo "ehlo $(hostname)"; sleep 1.5; echo "mail from: <root@$(hostname)>"; sleep 1.5; echo "quit"; sleep 1.5; } \
      | telnet ${SOURCE_ADDR:+"-b" "${SOURCE_ADDR}"} "$(echo "${HOST}" | awk '{print $2}')" 25 2>&1
  }

  # Run check_host() and grep its output into a variable for further use
  RESPONSE="$(check_host \
    | grep -Ei "block|blacklist|not allowed|banned|denied|rejected|ivmsip|invaluement|sorbs|spamcop|spamhaus")"

  # If RESPONSE had a match, we're blocked
  if [[ -n "${RESPONSE}" ]]; then
    IS_BLOCKED='true'
    RESULT="${TEXT_RED}FAIL${TEXT_RESET}"; else
    RESULT="${TEXT_GREEN}OK${TEXT_RESET}"
  fi

  # Display the host and its result (OK or FAIL)
  echo -e "${TEXT_BOLD}* $(echo "${HOST}" | awk -F "," '{print $1}') [${RESULT}${TEXT_BOLD}]${TEXT_RESET}"

  # For blocked results, we need to also show the error message
  if [[ -n "${RESPONSE}" ]]; then
    echo "🚫 ${RESPONSE})"
  fi

  # Setup for the Verizon/Yahoo check further up
  if [[ "${HOST}" == "Aol"* ]]; then
    AOL_RESPONSE="${RESPONSE}"
  fi

  # Same for the Outlook check
  if [[ "${HOST}" == "Hotmail"* ]]; then
    HOTMAIL_RESPONSE="${RESPONSE}"
  fi
done

# Final summary message and cleanup for telcheck.min
if [[ -v IS_BLOCKED ]]; then
  echo -e "\nBlock(s) detected. Follow the steps in the above output(s) to delist."
  unset IS_BLOCKED; else
  echo -e "\nAll clear! No blocks detected."
fi

# Unset SOURCE_ADDR for telcheck.min
if [[ -v SOURCE_ADDR ]]; then
  unset SOURCE_ADDR
fi
